src/viam/sdk/robot/client.cpp

```diff
@@ -57,6 +53,101 @@ using viam::robot::v1::Status;
+    query.model = proto.model();
+    return query;
+}
+
+RobotClient::discovery from_proto(const Discovery& proto) {
+    RobotClient::discovery discovery;
+    discovery.query = from_proto(proto.query());
+    discovery.results = struct_to_map(proto.results());
+    return discovery;
+}
+
+RobotClient::frame_system_config from_proto(const FrameSystemConfig& proto) {
+    RobotClient::frame_system_config fsconfig;
+    fsconfig.frame = WorldState::transform::from_proto(proto.frame());
+    if (proto.has_kinematics()) {
+        fsconfig.kinematics = struct_to_map(proto.kinematics());
+    }
+    return fsconfig;
+}
+
+RobotClient::status from_proto(const Status& proto) {
+    RobotClient::status status;
+    if (proto.has_name()) {
+        status.name = Name::from_proto(proto.name());
+    }
+    if (proto.has_status()) {
+        status.status_map = struct_to_map(proto.status());
+    }
+    if (proto.has_last_reconfigured()) {
+        status.last_reconfigured = timestamp_to_time_pt(proto.last_reconfigured());
+    }
+    return status;
+}
+
+RobotClient::operation from_proto(const Operation& proto) {
+    RobotClient::operation op;
+    op.id = proto.id();
+    op.method = proto.method();
+    if (proto.has_session_id()) {
+        op.session_id = proto.session_id();
+    }
+    if (proto.has_arguments()) {
+        op.arguments = struct_to_map(proto.arguments());
+    }
+    if (proto.has_started()) {
+        op.started = timestamp_to_time_pt(proto.started());
+    }
+    return op;
+}
+}  // namespace
+
+bool operator==(const RobotClient::discovery_query& lhs, const RobotClient::discovery_query& rhs) {
+    return lhs.subtype == rhs.subtype && lhs.model == rhs.model;
+}
+
+bool operator==(const RobotClient::discovery& lhs, const RobotClient::discovery& rhs) {
+    return lhs.query == rhs.query && map_to_struct(lhs.results).SerializeAsString() ==
+                                         map_to_struct(rhs.results).SerializeAsString();
+}
+
+bool operator==(const RobotClient::frame_system_config& lhs,
+                const RobotClient::frame_system_config& rhs) {
+    return lhs.frame == rhs.frame && map_to_struct(lhs.kinematics).SerializeAsString() ==
+                                         map_to_struct(rhs.kinematics).SerializeAsString();
+}
+
+bool operator==(const RobotClient::status& lhs, const RobotClient::status& rhs) {
+    return lhs.name == rhs.name &&
+           map_to_struct(lhs.status_map).SerializeAsString() ==
+               map_to_struct(rhs.status_map).SerializeAsString() &&
+           lhs.last_reconfigured == rhs.last_reconfigured;
+}
+
+bool operator==(const RobotClient::operation& lhs, const RobotClient::operation& rhs) {
+    return lhs.id == rhs.id && lhs.method == rhs.method && lhs.session_id == rhs.session_id &&
+           lhs.arguments == rhs.arguments && lhs.started == rhs.started;
+}
+
+struct RobotClient::impl {
+    impl(std::unique_ptr<RobotService::Stub> stub) : stub_(std::move(stub)) {}
```

Comment:
I still think you can write this as `impl(RobotService::Stub&& stub) : stub_(std::move(stub)` and get rid of the `unique_ptr`, but, at the same time, what you have here isn't wrong, it just has an extra level of indirection.

Reply:
fwiw I get errors when I try to make this change since the `RobotService::NewStub` method returns a `unique_ptr` and not a raw `Stub` type. I can get rid of the `unique_ptr` by dereferencing it in the constructor but I'm not sure that's better.

Reply:
I tried this twice before and it gave me errors even though I changed everything. It's a little hard to find the two conversations we had on this since I was asked not to resolve the comments.
I'm guessing it's because of what Ethan just said.